### ISSUE TITLE
How to call sgx_ra_get_msg1?

### ISSUE THREAD

**Issue by ech0-de** [apache/teaclave-sgx-sdk#68]
I can't figure out how to correctly call `sgx_ra_get_msg1`, because I cannot convince Rust to use the `sgx_ra_get_ga` function pointer.

So far I have added the following two lines to my Enclave.edl, so that edger8r adds `sgx_ra_get_ga` to Enclave_u.c and Enclave_u.h:
```
include "sgx_tkey_exchange.h"
from "sgx_tkey_exchange.edl" import *;
```

However, the Rust compiler is still unable to find `sgx_ra_get_ga`, so I continued by adding the following lines to my `app/src/main.rs` (that also contains the `sgx_ra_get_msg1` call):
```
extern {
    fn sgx_ra_get_ga(eid: sgx_enclave_id_t, retval: *mut sgx_status_t,
                               context: sgx_ra_context_t, g_a: *mut sgx_ec256_public_t) -> sgx_status_t;
}
```

With this snippet in place the compiler fails with the following error message:
```
error[E0308]: mismatched types
   --> src/main.rs:153:52
    |153 |         sgx_ra_get_msg1(context, enclave.geteid(), sgx_ra_get_ga, &mut msg1)
    |                                                    ^^^^^^^^^^^^^ expected normal fn, found unsafe fn
    |
    = note: expected type `extern "C" fn(u64, *mut sgx_types::sgx_status_t, u32, *mut sgx_types::sgx_ec256_public_t) -> sgx_types::sgx_status_t`
               found type `unsafe extern "C" fn(u64, *mut sgx_types::sgx_status_t, u32, *mut sgx_types::sgx_ec256_public_t) -> sgx_types::sgx_status_t {sgx_ra_get_ga}
```

I doubt that both of the above steps are actually necessary, but I'm completely clueless how to call the function correctly. Unfortunately for me, the only samplecode that uses the `sgx_ra_get_msg1` is the remoteattestation example, which is written in C++ and thus rather different.

Thanks!

**Comment by dingelish** [comment]
Looking into this problem. Thanks for your report!

**Comment by dingelish** [comment]
Seems that your solution is almost correct except for using `unsafe {...}` to make calls to `sgx_ra_get_ga`.

try this:
```
unsafe { sgx_ra_get_msg1(context, enclave.geteid(), sgx_ra_get_ga, &mut msg1) }
```

**Comment by dingelish** [comment]
`sgx_ra_get_ga` is not an official SGX API. The definition is [here](https://github.com/intel/linux-sgx/blob/master/sdk/tkey_exchange/tkey_exchange.cpp#L92). Don't know if we should provide it in sgx_types as an extern function. What do you think?

**Comment by ech0-de** [comment]
Thanks for the pointer in the right direction. I now managed to get it to work, by adding the `sgx_ra_get_ga` to sgx_types and also modifying the existing type `sgx_ecall_get_ga_trusted_t` by adding an `unsafe` before the `extern "c"`.

My minimum working example looks something like this:
```rust
    let mut msg1 = sgx_ra_msg1_t::default();
    let mut context = sgx_ra_context_t::default();

    unsafe {
        ra_create_context(enclave.geteid(), &mut retval, &mut context);
        sgx_ra_get_msg1(context, enclave.geteid(), sgx_ra_get_ga, &mut msg1);
    };
```
If anyone else is trying to do this, note that this also requires linking against `sgx_ukey_exchange`, by adding it to `app/build.rs`.

From my understanding `sgx_ra_get_msg1` (and related functions) are Intel's recommended way to perform a remote attestation using Intel IAS as described [here](https://software.intel.com/en-us/articles/code-sample-intel-software-guard-extensions-remote-attestation-end-to-end-example). Since this method requires requires a function pointer to `sgx_ra_get_ga`, I would suggest to provide it as an extern function in sgx_types. But I'm not entirely sure if that is really the correct move.

Here are my changes to `sgx_types/src/function.rs` and `sgx_types/src/types.rs`:
```diff
diff --git a/sgx_types/src/function.rs b/sgx_types/src/function.rs
index 7c767bb..91869f7 100644
--- a/sgx_types/src/function.rs
+++ b/sgx_types/src/function.rs
@@ -396,6 +396,9 @@ extern {
                            p_key: * mut sgx_ra_key_128_t) -> sgx_status_t;
 
     pub fn sgx_ra_close(context: sgx_ra_context_t) -> sgx_status_t;
+
+    pub fn sgx_ra_get_ga(eid: sgx_enclave_id_t, retval: *mut sgx_status_t,
+                         context: sgx_ra_context_t, g_a: *mut sgx_ec256_public_t) -> sgx_status_t;
 }
```
```diff
diff --git a/sgx_types/src/types.rs b/sgx_types/src/types.rs
index 7b64284..6c95d20 100644
--- a/sgx_types/src/types.rs
+++ b/sgx_types/src/types.rs
@@ -967,7 +967,7 @@ impl_struct! {
 //
 
 
-pub type sgx_ecall_get_ga_trusted_t = extern "C" fn(eid: sgx_enclave_id_t,
+pub type sgx_ecall_get_ga_trusted_t = unsafe extern "C" fn(eid: sgx_enclave_id_t,
                                                     retval: * mut sgx_status_t,
                                                     context: sgx_ra_context_t,
                                                     g_a: * mut sgx_ec256_public_t) -> sgx_status_t;
```

If you think that these changes make sense, I can of course create a pull request for them.

**Comment by dingelish** [comment]
I think you are right! I'd happy to merge your PR. Thanks!

**Comment by dingelish** [comment]
Fixed in dd6e4a2c98b2011a13812bf9e7e7d9a8614928c5
Thanks!
